interceptor.go

```diff
@@ -0,0 +1,90 @@
+// +build !js
+
+package webrtc
+
+import (
+	"io"
+
+	"github.com/pion/rtcp"
+	"github.com/pion/rtp"
+)
+
+// WriteRTP is used by Interceptor.BindLocalTrack.
+type WriteRTP func(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error)
```

Comment:
I think it's better to define Attribute type in the same way as `net/http.Header`.
Does it require to take `interface{}` as a key instead of `string`?

Reply:
Personally, I like to make it Reader/Writer interface as like `io` stdlib.
For example
```go
type WriterRTP interface {
  Write(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error)
}
```
It may have an adapter like
```go
type WriterRTPFunc func(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error)
func (f WriterRTPFunc) Write(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error) {
  return f(p, attribute)
}
```

Reply:
Oh yea that is a good point, forgot about that! `io.Reader` is nice because it integrates into other things easily.

If we satisfy that interface then we can use a bunch of existing stuff.

Reply:
That can be implemented right now with the `TrackLocalWriter`

Reply:
For me is not very clear the purpose of the attributes map

Reply:
For me personally, I really like http.ResponseWriter approach, add a method to WriterRTP to retrieve the attributes, and the retrieved attributes will have its own type that contains setters/getters:

```golang
type Attributes map[interface{}]interface{}
func (attributes Attributes) Get(k interface{}) interface{} {}
func (attributes Attributes) Set(k, v interface{}) {}
func (attributes Attributes) GetUint32(k interface{}) uint32 {}
func (attributes Attributes) SetUint32(k, v uint32) {}

type WriterRTP interface {
  Write(p *rtp.Packet) (int, error)
  Attributes() Attributes
}
```

I think this approach gives us some benefits:
  1. By having Attributes type, it allows us to change its implementation in the future, as we can operate on the Get/Set methods, not touching the map directly.
  2. Since we can add methods to Attributes, it's possible to add static types setter/getter, e.g. GetInt32/SetInt32, GetFloat32/SetFloat32, etc. Although we still store these values as interface{}, at least we can hide the type casting inside of Attributes, which I think will make the caller simpler and less error prone.

For the adapter, we can have it as:

```golang
type writerRTPAdapter struct {
  Write func(p *rtp.Packet) (int, error)
  Attributes Attributes
}

func (adapter *writerRTPAdapter) Write(p *rtp.Packet) (int, error) {}
func (adapter *writerRTPAdapter) Attributes() Attributes {}
```

Reply:
I'm very happy the attributes handling got so much attention.

## Purpose of the attributes

The purpose of attributes is to pass metadata belonging to the packet between interceptors, and ideally between interceptor and the user (developer). Some idea for the usage:
- attaching arrival time to packets in the first interceptor (because some interceptors, like a jitter buffer will delay packets, so calling time.Now after those will not return the correct arrival time)
- calculating an unwrapped sequence number once, and passing it
- adding a send time time.Time attribute to packets, for audio/video sync (this can be done using the NTP time from incoming rtcp SenderReports) - this would be most useful if the user could read the attributes too, not just the interceptors
- debugging

## Passing attributes with rtp packets

I'm really not happy with how attributes are passed right now. I think it would be best to attach them to the RTP packet somehow, I just haven't found a good solution to that. Some things I considered:
- adding Attributes field to rtp.Packet
    - this would modify the public api
- adding attributes unexported field to rtp.Packet, and adding a GetAttribute(p *Packet, key interface{}) interface{}, and a similar SetAttribute method to the rtp packet
    - this is ugly
- use a separate global map for attributes (eg: key is ssrc+sequence number, value is attributes)
    - it's hard to know when it's safe to remove attributes for a packet
    - also very ugly
- modifying rtp.Packet to be an interface, and creating a wrapper for this in interceptor packet(see below)
   - this would require a LOT of modification
```
type PacketWithAttribute struct {
    rtp.Packet
    attributes map[interface{}]interface{}
}

func GetPacketAttribute(p rtp.Packet, key interface{}) interface{} {
    pa, ok := p.(*PacketWithAttribute)
    if !ok {
        return nil
    }
    return pa.attributes[key]
} 
```

Any idea is welcome for this!

## Attributes type and key type
The key type is interface{} instead of string, because I planned to use attributes with unexported struct keys, similarly to how context.Context usage is recommended (this would also mean generic SetUint32 and similar methods would be unnecesary, as every attribute would have it's own setter/getter):
```
    type fooKey struct{}

    func SetFoo(attributes map[interface{}]interface{}, value string) {
        attributes[fooKey{}] = value
    }

    func GetFoo(attributes map[interface{}]interface{}) string {
        str, _ := attributes[fooKey{}].(string)
        return str
    }
```

I'm all for using a dedicated `type Attribtes map[interface{}]interface{} though. We can simply use string keys as well if you guys think this is unnecessarily complicated.

## Function vs Writer interface
@at-wat 
I initially chose function types vs interfaces so the user don't need to define a lot of structs. But this adapter function is awesome, never saw this pattern before :)

```
type WriterRTPFunc func(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error)
func (f WriterRTPFunc) Write(p *rtp.Packet, attributes map[interface{}]interface{}) (int, error) {
  return f(p, attribute)
}
```

This is a great idea, I will do this!

## http.ResponseWriter approach
@lherman-cs 
The equivalent of a http response is an rtp packet. http.ResponseWriter writes the body and headers of a single http response. That means sadly we can't use the same aproach for writing attributes in RTPWriter as is used for writing headers in ResponseWriter, because RTPWriter writes multiple rtp packets, each with different attributes, while http.ResponseWriter writes a single http response.

Reply:
@masterada Now is clear for me the usage of attributes, but would not be better to use `context.Context`? or there is a special reason to use an `interface` map?

Reply:
There are 3 reasons why I used map instead of context:
1. In my original proposal context.Context was passed in Read/Write functions, so it could be used for context deadiline/context cancel. But the read/write call in sender/receiver do not support these, so I dropped them to avoid confusion.
2. In the read methods, packets and attributes are returned. As context.Context is usually passed as to functions and not returned, I would find it confusing to use them as return values. It would be even more confusing if in the future we would add context parameter to read for deadlin/cancel, because that would mean a context is used as an argument, and another context is retruned, and the two would have very different purpose.
3. Adding a value to the context needs wrapping, and getting it means a recursive call. This feels like this could cause some performance issues in some cases, however I didn't do any benchmarks so this is just a hunch.